src/controllers/EvidenceUploadController.ts

```diff
@@ -24,36 +32,100 @@ const navigation = {
+
+        return await this.renderWithStatus(UNPROCESSABLE_ENTITY)(
+            this.template, {
+                ...this.prepareViewModelFromAppeal(appeal),
+                ...this.httpContext.request.body,
+                validationResult
+            }
+        );
+    }
+
     protected getExtraActionHandlers(): Record<string, FormActionHandler | FormActionHandlerConstructor> {
         const that = this;
+
+        const fileTooLargeError: string = 'File size must be smaller than 4MB';
+        const fileNotSupportedError: string = 'The selected file must be a TXT, DOC, PDF, JPEG or PNG';
+        const tooManyFilesError: string = 'You can only select up to 10 files at the same time';
+
         return {
             'upload-file': {
                 async handle(request: Request, response: Response): Promise<void> {
-                    await parseFormData(request, response);
-
-                    if (!request.file) {
-                        response.redirect(request.route.path);
-                        return;
-                    }
-
-                    const id = await that.fileTransferService.upload(request.file.buffer, request.file.originalname);
 
                     const appeal: Appeal = request.session
                         .chain(_ => _.getExtraData())
                         .map<ApplicationData>(data => data[APPLICATION_DATA_KEY])
                         .map(data => data.appeal)
                         .unsafeCoerce();
 
+                    try {
+                        await parseFormData(request, response)
+                    } catch (error) {
+                        switch (error.code) {
+                            case 'LIMIT_FILE_SIZE':
+                                return await that.renderUploadError(appeal, fileTooLargeError);
+                            case 'LIMIT_UNEXPECTED_FILE':
+                                return await that.renderUploadError(appeal, fileNotSupportedError);
+                        }
+                    }
+
+                    if (!request.file) {
+                        response.redirect(MOVED_TEMPORARILY, request.route.path);
+                        return;
+                    } else if (appeal.reasons.other.attachments &&
+                        appeal.reasons.other.attachments!.length >= maxNumberOfFiles) {
+                        return await that
+                            .renderUploadError(appeal, tooManyFilesError);
+                    }
+
+                    let id: string;
+
+                    try {
+                        id = await that.fileTransferService.upload(request.file.buffer, request.file.originalname);
+                    } catch (err) {
+                        if (err.message === 'Unsupported file type') {
+                            return await that
+                                .renderUploadError(appeal, fileNotSupportedError);
+                        } else {
+                            throw new Error(err.message)
```

Comment:
I guess you can just re-throw error e.g. `throw err`. I see no value in wrapping it in another error.